src/wallet/test/walletdb_tests.cpp

```diff
@@ -29,5 +29,98 @@ BOOST_AUTO_TEST_CASE(walletdb_readkeyvalue)
     BOOST_CHECK_THROW(ssValue >> dummy, std::ios_base::failure);
 }
 
+BOOST_AUTO_TEST_CASE(walletdb_readonly_database)
+{
+    // Test read-only database functionality including property detection,
+    // batch operations, and LoadWallet behavior
+
+    // Test 1: IsReadOnly() property detection
+    // Regular database should not be read-only
+    auto regular_db = CreateMockableWalletDatabase();
+    BOOST_CHECK(!regular_db->IsReadOnly());
+
+    // Read-only database should report as read-only
+    auto readonly_db = CreateMockableWalletDatabase({}, true);
+    BOOST_CHECK(readonly_db->IsReadOnly());
+
+    // Test with wallet instances
+    std::shared_ptr<CWallet> regular_wallet = std::make_shared<CWallet>(nullptr, "", std::move(regular_db));
+    BOOST_CHECK(!regular_wallet->GetDatabase().IsReadOnly());
+
+    std::shared_ptr<CWallet> readonly_wallet = std::make_shared<CWallet>(nullptr, "", std::move(readonly_db));
+    BOOST_CHECK(readonly_wallet->GetDatabase().IsReadOnly());
+
+    // Test 2: Batch operations with read-only database
+    MockableData initial_data;
+
+    // Pre-populate with some wallet-like data
+    DataStream key_stream, value_stream;
+    key_stream << std::make_pair(DBKeys::VERSION, std::string{});
+    value_stream << CLIENT_VERSION;
+    initial_data[SerializeData(key_stream.begin(), key_stream.end())] = SerializeData(value_stream.begin(), value_stream.end());
+
+    key_stream.clear();
+    value_stream.clear();
+    key_stream << std::make_pair(DBKeys::FLAGS, std::string{});
+    value_stream << uint64_t{WALLET_FLAG_DESCRIPTORS};
+    initial_data[SerializeData(key_stream.begin(), key_stream.end())] = SerializeData(value_stream.begin(), value_stream.end());
+
+    auto readonly_database = CreateMockableWalletDatabase(initial_data, true);
+
+    // Test that we can create a batch and perform read operations
+    std::unique_ptr<DatabaseBatch> batch = readonly_database->MakeBatch();
+
+    // Reading should work
+    int version;
+    BOOST_CHECK(batch->Read(std::make_pair(DBKeys::VERSION, std::string{}), version));
+    BOOST_CHECK_EQUAL(version, CLIENT_VERSION);
+
+    uint64_t flags;
+    BOOST_CHECK(batch->Read(std::make_pair(DBKeys::FLAGS, std::string{}), flags));
+    BOOST_CHECK_EQUAL(flags, WALLET_FLAG_DESCRIPTORS);
```

Comment:
Same as my previous comment, this test is not useful as it only test `MockableDatabase` which is only used by tests. For such a test to be useful, it would need to be making a `WalletBatch` so that it is actually reaching code that will be used in production.

Reply:
So you mean I could make these tests useful by modifying `WalletBatch` but `MockableDatabase`? or I could just delete them?